### PR TITLE
WIP: SC Branch

### PR DESCRIPTION
**Pull request by ghalym**
## Description
This branch should include all the changes required for the PMPS library to support SC.
Among those changes are the inclusion of two function blocks to read a json file and parse it into the SafeBeam parameters structure.

## Pre-merge checklist
- [x] Libraries are set to ``Always Newest`` version (``Library, *``)
- [x] Code works interactively
- [x] Code contains descriptive comments
- [x] Test suite passes locally
- [ ] Committed with ``pre-commit`` or ran ``pre-commit run --all-files``
- [ ] Check that ``BP_IO`` parameters weren't modified unintentionally
image

### CODE REVIEW

File: lcls-twincat-pmps/PMPS/AncillaryComponents/FB_JsonDocToSafeBP.TcPOU [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+		ELSE
+			Step := 900 + Step;
+			nErrId := Step; 
+			bError := TRUE; 
+			sErrMsg:=  CONCAT('Error device name not found in Json File : ',sDeviceName);		
+		END_IF
+		
+	 21: (*Check device States in DOM*)
+	 		FOR Index:=1 TO MAX_STATES BY 1 DO
+				  IF NOT ((arrStates[Index].sPmpsState= '') OR (arrStates[Index].sPmpsState= 'Invalid')) THEN
```
Comment (ZLLentz) [inline lcls-twincat-pmps/PMPS/AncillaryComponents/FB_JsonDocToSafeBP.TcPOU:82]: These need to be `sName` or else the comparison doesn't work properly. `sPmpsState` will never be `Invalid`, for example.

Comment (ghalym) [reply lcls-twincat-pmps/PMPS/AncillaryComponents/FB_JsonDocToSafeBP.TcPOU]: Thank you! fixed in c5646af

Comment (ghalym) [reply lcls-twincat-pmps/PMPS/AncillaryComponents/FB_JsonDocToSafeBP.TcPOU]: Sorry to ask this again, What is the difference between sName and sPmpsState? Which one should be in the Json file?

Comment (ZLLentz) [reply lcls-twincat-pmps/PMPS/AncillaryComponents/FB_JsonDocToSafeBP.TcPOU]: `sName` is the human-readable short name such as "OUT" or "YAG"
`sPmpsState` is the lookup key such as "IM1K0-XTES-YAG"

Comment (ghalym) [reply lcls-twincat-pmps/PMPS/AncillaryComponents/FB_JsonDocToSafeBP.TcPOU]: So sPmpsState in the value that will be in the json file. So I should revert the changes and in step 21 check that sPmpsState is in the DOM.

Comment (ZLLentz) [reply lcls-twincat-pmps/PMPS/AncillaryComponents/FB_JsonDocToSafeBP.TcPOU]: The code I was running right before Thanksgiving seemed to do everything correctly in the json reading

File: lcls-twincat-pmps/PMPS/AncillaryComponents/FB_JsonDocToSafeBP.TcPOU [hunk 2]
```diff
@@ -0,0 +1,252 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<TcPlcObject Version="1.1.0.1" ProductVersion="3.1.4022.18">
+  <POU Name="FB_JsonDocToSafeBP" Id="{be21f50f-6e98-4ce2-92ee-0a1b1ced09d9}" SpecialFunc="None">
+    <Declaration><![CDATA[FUNCTION_BLOCK FB_JsonDocToSafeBP
+VAR_INPUT
+	bExecute             : BOOL; //Rising Edge
+	jsonDoc 		: SJsonValue;
+	sDeviceName : STRING;
+	//asDeviceStateNames: ARRAY [1.. MOTION_GVL.MAX_STATES] OF STRING;
```
Comment (ZLLentz) [inline lcls-twincat-pmps/PMPS/AncillaryComponents/FB_JsonDocToSafeBP.TcPOU:9]: We can probably remove this leftover comment

File: lcls-twincat-pmps/PMPS/AncillaryComponents/FB_JsonDocToSafeBP.TcPOU [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+	bHasDevice		: BOOL;
+	bHasAllStates		: BOOL;
+	bHasAllParameters :BOOL := TRUE;
+	bBusy                : BOOL;
+    bError               : BOOL;
+    nErrId               : UDINT;
+	sErrMsg				 :STRING;
+END_VAR
+VAR_IN_OUT
+	//ARRAY [1.. MOTION_GVL.MAX_STATES] ;
```
Comment (ZLLentz) [inline lcls-twincat-pmps/PMPS/AncillaryComponents/FB_JsonDocToSafeBP.TcPOU:21]: leftover comment

File: lcls-twincat-pmps/PMPS/AncillaryComponents/FB_JsonDocToSafeBP.TcPOU [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+		ELSE
+			Step := 900 + Step;
+			nErrId := Step; 
+			bError := TRUE; 
+			sErrMsg:=  CONCAT('Error device name not found in Json File : ',sDeviceName);		
+		END_IF
+		
+	 21: (*Check device States in DOM*)
+	 		FOR Index:=LOWER_BOUND(arrStates,1) TO UPPER_BOUND(arrStates,1) BY 1 DO
+				  IF NOT ((arrStates[Index].sPmpsState= '') OR (arrStates[Index].sPmpsState= 'Invalid')) THEN
```
Comment (ZLLentz) [inline lcls-twincat-pmps/PMPS/AncillaryComponents/FB_JsonDocToSafeBP.TcPOU:82]: I don't think `sPmpsState` will ever be `'Invalid'`, I think this used to be an `sName` check. I don't think this causes any issues though.

File: lcls-twincat-pmps/PMPS/GVLs/PMPS_GVL.TcGVL [hunk 5]
```diff
@@ -62,6 +75,7 @@ VAR_GLOBAL CONSTANT
         nTran := 1,
         neVRange :=  2#1111_1111_1111_1111_1111_1111_1111_1111,
         nRate    := UDINT#1000000,
+		nBCRange := 2#0111_1111_1111_1111,
```
Comment (ZLLentz) [inline lcls-twincat-pmps/PMPS/GVLs/PMPS_GVL.TcGVL:78]: If the "spare" bps end up being bad, we can switch this to `2#0001_1111_1111_1111`

File: lcls-twincat-pmps/PMPS/HelperFunctions/F_DifferentBeamParams.TcPOU [hunk 6]
```diff
@@ -26,6 +28,8 @@ END_VAR
       <ST><![CDATA[xAttOK := BeamParam1.nTran <> BeamParam2.nTran;
 xEvOK := BeamParam1.neVRange <> BeamParam2.neVRange;
 xRateOK     := BeamParam1.nRate <> BeamParam2.nRate;
+xBCOK	:= BeamParam1.nBCRange <> BeamParam2.nBCRange;// OR BeamParam1.nBeamClass <> BeamParam2.nBeamClass ; //Is this right?
```
Comment (ZLLentz) [inline lcls-twincat-pmps/PMPS/HelperFunctions/F_DifferentBeamParams.TcPOU:31]: I think this is correct as-is and doesn't need the OR (responding to the comment)

File: lcls-twincat-pmps/PMPS/HelperFunctions/F_SafeBPCompare0Rate.TcPOU [hunk 7]
```diff
@@ -3,24 +3,46 @@
...
 VAR
-    // Beam-rate is zero, masks all others considerations.
+    {attribute 'pytmc' := '
+        pv: LFE:LCLSMODE
+        link: SIOC:FEEH:MP01:LCLSMODE_RBV
+    '}
+    HXR_MODE : FB_LREALFromEPICS;
+	{attribute 'pytmc' := '
+        pv: KFE:LCLSMODE
+        link: SIOC:FEES:MP01:LCLSMODE_RBV
```
Comment (ZLLentz) [inline lcls-twincat-pmps/PMPS/HelperFunctions/F_SafeBPCompare0Rate.TcPOU:23]: Question: how does this behave if the PV here doesn't connect?

File: lcls-twincat-pmps/PMPS/MajorComponents/BPTM/BeamParameterTransitionManager.TcPOU [hunk 8]
```diff
@@ -8,6 +8,7 @@ NOTE:
 The BPTM will throw an error if the arbiter does not have enough space for the transition and new final assertion. 
 
  *)
+{attribute 'no_check'}
```
Comment (ZLLentz) [inline lcls-twincat-pmps/PMPS/MajorComponents/BPTM/BeamParameterTransitionManager.TcPOU:11]: Should we revisit these "no check" attributes later? It's a little bit concerning that they were all needed.

File: lcls-twincat-pmps/PMPS/MajorComponents/PMPS_MPS_Interface/FB_BeamClassWatcher.TcPOU [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+    <Implementation>
+      <ST><![CDATA[IF bInit THEN
+    fbGetHN(bExecute:=TRUE);
+    bInit R= NOT fbGetHN.bBusy;
+END_IF
+    
+xBeamClassWithinBounds := (i_stCurrentBeamParams.nBCRange AND i_stRequestedBeamParams.nBCRange) = i_stCurrentBeamParams.nBCRange
+					AND  (i_stCurrentBeamParams.nBeamClass <= i_stRequestedBeamParams.nBeamClass);
+
+bcResidual := (i_stCurrentBeamParams.nBCRange XOR i_stRequestedBeamParams.nBCRange) AND i_stCurrentBeamParams.nBCRange;
```
Comment (ZLLentz) [inline lcls-twincat-pmps/PMPS/MajorComponents/PMPS_MPS_Interface/FB_BeamClassWatcher.TcPOU:75]: Do you want this residual displayed in the ui?

File: lcls-twincat-pmps/PMPS/PMPS.plcproj [hunk 10]
```diff
@@ -483,22 +510,22 @@
   </ItemGroup>
   <ItemGroup>
     <PlaceholderResolution Include="LCLS General">
-      <Resolution>LCLS General, * (SLAC)</Resolution>
+      <Resolution>LCLS General, 0.0.0 (SLAC)</Resolution>
```
Comment (ZLLentz) [inline lcls-twincat-pmps/PMPS/PMPS.plcproj:513]: Do we need to release general lib and switch this back?

Comment (ZLLentz) [reply lcls-twincat-pmps/PMPS/PMPS.plcproj:513]: I think this is the only thing that really needs fixing in the PR

**Review by ZLLentz — COMMENTED** [review]
Hi Maggie, this is 99% good to me! I agree with the design decisions and it works in practice.

I found one specific thing that needs fixing (the lib uses general version 0.0.0) and everything else is less important.

File: lcls-twincat-pmps/PMPS/PMPS.plcproj (hunk 10 above)
Comment (ghalym) [reply lcls-twincat-pmps/PMPS/PMPS.plcproj:513]: Thanks for all your comments Zach. I will work on addressing them soon before we start with the KFE deployment.

File: lcls-twincat-pmps/PMPS/AncillaryComponents/FB_JsonDocToSafeBP.TcPOU [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+			  Step := 900+ Step;
+			  nErrId := Step; 
+			  bError := TRUE; 
+			  sErrMsg:= CONCAT('Error Input State array is empty : ',sDeviceName);				  
+		 ELSE 
+			 Step := Step+1;		 
+		 END_IF;
+
+	 23:  (*loop Device State Arrays to find all States*)
+	      FOR Index:=1 TO nStateCount BY 1 DO
```
Comment (ZLLentz) [inline lcls-twincat-pmps/PMPS/AncillaryComponents/FB_JsonDocToSafeBP.TcPOU:111]: The logic in this section is broken for arrays that don't start at 1 (my current error) and arrays that end at 1 (hypothesis)
I picked array position 0 for the transition states and they aren't loading

File: lcls-twincat-pmps/PMPS/GVLs/PMPS_GVL.TcGVL (hunk 5 above)
Comment (ZLLentz) [reply lcls-twincat-pmps/PMPS/GVLs/PMPS_GVL.TcGVL:68]: spare bps are ok so I can mark this comment as resolved

File: lcls-twincat-pmps/PMPS/HelperFunctions/F_SafeBPCompare.TcPOU [hunk 12]
```diff
@@ -45,7 +48,7 @@ FOR idx:=1 to PMPS_GVL.MAX_APERTURES DO
     BeamParam1.astApertures[idx].Width > BeamParam2.astApertures[idx].Width;
 END_FOR
 
-F_SafeBPCompare := xAttOK AND xEvOK AND xRateOK AND xAuxAprtOK AND xAuxAttOK;
+F_SafeBPCompare := xAttOK AND xEvOK AND xRateOK AND xAuxAprtOK AND xAuxAttOK AND xBCOK;
```
Comment (ZLLentz) [inline lcls-twincat-pmps/PMPS/HelperFunctions/F_SafeBPCompare.TcPOU:51]: Random thought: in the event that both rate and bc were ok, could we ignore the machine mode being invalid?
Might be unnecessarily complicated but it's a thought

Comment (ZLLentz) [reply lcls-twincat-pmps/PMPS/HelperFunctions/F_SafeBPCompare.TcPOU:51]: Same for the zero rate compare version

**Review by ZLLentz — APPROVED** [review]
I think my old comments still apply but everything is minor or a question or something like that.
I'm comfortable running the branch as-is on the component PLCs (assuming all commits have been pushed)

File: lcls-twincat-pmps/PMPS/HelperFunctions/F_SafeBPCompare0Rate.TcPOU (hunk 7 above)
Comment (ghalym) [reply lcls-twincat-pmps/PMPS/HelperFunctions/F_SafeBPCompare0Rate.TcPOU]: Should Fault the beam

**Pull request merged**